### PR TITLE
AP_OSD: imperial units and warning levels

### PR DESCRIPTION
**Pull request by sh83**
- Added support for imperial units
- Fixed lat/lon field width, so it will display well for 3 digit number with sign
- Add warning levels for vbat, rssi, nsat
- Add setting for inverted roll AHI and inverted wind direction
- Added fine tune settings to adjust osd inside screen (using VOS and HOS MAX7456 registers)
- Reworked default osd items position ( thanks to @Kelly-Foster  for the layout)
- Fixed spi buffer size calculation

### CODE REVIEW

File: libraries/AP_OSD/AP_OSD.cpp [hunk 1]
```diff
@@ -73,10 +73,45 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
...
+
+    // @Param: _W_RSSI
+    // @DisplayName: RSSI warn level (in %)
+    // @Description: Set level at which RSSI item will flash
+    // @Range: 0 99
+    // @User: Standard
+    AP_GROUPINFO("_W_RSSI", 11, AP_OSD, warn_rssi, 30),
+
+    // @Param: _W_NSAT
+    // @DisplayName: NSAT warn leve
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD.cpp]: spelling of "level"

File: libraries/AP_OSD/AP_OSD.cpp [hunk 2]
```diff
@@ -140,6 +175,7 @@ void AP_OSD::update_osd()
 
     update_current_screen();
 
+    screen[current_screen].set_osd(this);
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD.cpp]: what we normally do is add an accessor on the backend class. So in this case add get_options() to the AP_OSD_Backend.h

File: libraries/AP_OSD/AP_OSD.h [hunk 3]
```diff
@@ -162,10 +169,20 @@ class AP_OSD {
     AP_Int8 rc_channel;
     AP_Int8 sw_method;
 
+    AP_Int8 v_offset;
+    AP_Int8 h_offset;
+
+    AP_Int8 warn_rssi;
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD.h]: I think the warn_rssi better be a AP_Int16, as levels over 127 may be wanted

File: libraries/AP_OSD/AP_OSD.h [hunk 4]
```diff
@@ -162,10 +169,20 @@ class AP_OSD {
...
+    AP_Int8 warn_nsat;
+    AP_Float warn_batvolt;
+
     enum {
         OPTION_DECIMAL_PACK = 1U<<0,
+        OPTION_INVERTED_WIND = 1U<<1,
+        OPTION_INVERTED_AH_ROLL = 1U<<2,
+        OPTION_IMPERIAL_UNITS = 1U<<3,
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD.h]: I think we'll need a separate OSD_UNITS parameter, as I'm sure we will end up with several possible units systems (eg. knots, m/s, mph, km/hr etc)

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 5]
```diff
@@ -313,7 +333,11 @@ void AP_OSD_Screen::draw_speed_vector(uint8_t x, uint8_t y,Vector2f v, int32_t y
         int32_t interval = 36000 / SYM_ARROW_COUNT;
         arrow = SYM_ARROW_START + ((angle + interval / 2) / interval) % SYM_ARROW_COUNT;
     }
-    backend->write(x, y, false, "%c%3.0f%c", arrow,  v_length * 3.6, SYM_KMH);
+    if (osd->options.get() & AP_OSD::OPTION_IMPERIAL_UNITS) {
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD_Screen.cpp]: I think we need a helper function for writing with units, something like:
  write_units(char prefix_char, const char *fmt, float value);

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 6]
```diff
@@ -331,6 +355,11 @@ void AP_OSD_Screen::draw_horizon(uint8_t x, uint8_t y)
     float roll = ahrs.roll;
     float pitch = -ahrs.pitch;
 
+    //inverted roll AH
+    if (osd->options.get() & AP_OSD::OPTION_INVERTED_AH_ROLL) {
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD_Screen.cpp]: this would become if (get_options() & AP_OSD::OPTION_INVERTED_AH_ROLL)

Comment (tridge) [reply libraries/AP_OSD/AP_OSD_Screen.cpp]: or maybe a check_option() function that returns a bool?

File: libraries/AP_OSD/AP_OSD.h (hunk 3 above)
Comment (sh83) [reply libraries/AP_OSD/AP_OSD.h]: In osd rssi is shown in %, so 99 is the maximum level

File: libraries/AP_OSD/AP_OSD_Screen.cpp (hunk 6 above)
Comment (sh83) [reply libraries/AP_OSD/AP_OSD_Screen.cpp]: Thanks, good idea. Implemented.

File: libraries/AP_OSD/AP_OSD.h (hunk 4 above)
Comment (sh83) [reply libraries/AP_OSD/AP_OSD.h]: Added parameter.

File: libraries/AP_OSD/AP_OSD_Screen.cpp (hunk 5 above)
Comment (sh83) [reply libraries/AP_OSD/AP_OSD_Screen.cpp]: Added function to convert values based on chosen units system.

**Comment by tridge** [comment]
I've pushed some tidy-ups to this PR
I've also got SI and aviation units ready, pending symbols for Nm and Knots

**Pull request merged**